_posts/2019-01-03-elixir-kaffe-codealong.md

```diff
@@ -0,0 +1,204 @@
+    # straight from our machine. However, if the server is dockerized, the hostname will
+    # be called whatever is specified by that container (usually "kafka")
+    topics: ["our_topic", "another_topic"], # add a list of topics you plan to produce messages to
+  ]
+```
+
+**4. Configure consumer**
+**4.a:** add `/lib/application.ex` with the following code:
+```elixir
+defmodule ElixirKaffeCodealong.Application do
+  use Application # read more about Elixir's Application module here: https://hexdocs.pm/elixir/Application.html
+
+  def start(_type, args) do
+    import Supervisor.Spec
+    children = [
+      worker(Kaffe.Consumer, []) # calls to start Kaffe's Consumer module
+    ]
+    opts = [strategy: :one_for_one, name: ExampleConsumer.Supervisor]
+    Supervisor.start_link(children, opts)
+  end
+end
+```
+**4.b:** back in `mix.exs`, add a new item to the application function:
+```elixir
+def application do
+  [
+    extra_applications: [:logger, :kaffe],
+    mod: {ElixirKaffeCodealong.Application, []}
+    # now that we're using the Application module, this is where we'll tell it to start.
+    # We use the keyword `mod` with applications that start a supervision tree,
+    # which we configured when adding our Kaffe.Consumer to Application above.
+  ]
+end
+```
+**4.c:** add a consumer module to accept messages from Kafka as `/lib/example_consumer.ex` with the following code:
+```elixir
+defmodule ExampleConsumer do
+  # function to accept Kafka messaged MUST be named "handle_message"
+  # MUST accept arugments structured as shown here
+  # MUST return :ok
+  # Can do anything else within the function with the incoming message
+
+  def handle_message(%{key: key, value: value} = message) do # necessary
+    IO.inspect message # optional
+    IO.puts "#{key}: #{value}" # optional
+    :ok # necessary
+  end
+end
+```
+**4.d:** configure the consumer module in `/config/config.exs`
+```elixir
+config :kaffe,
+  consumer: [
+    endpoints: [localhost: 9092],               
+    topics: ["our_topic", "another_topic"],     # the topic(s) that will be consumed
+    consumer_group: "example-consumer-group",   # the consumer group for tracking offsets in Kafka
+    message_handler: ExampleConsumer,           # the module that will process messages
+  ]
+```
+
+**5. Add a producer module (optional, can also call Kaffe from the console)**
+add `/lib/example_producer.ex` with the following code:
+```elixir
+defmodule ExampleProducer do
+  def send_my_message({key, value}, topic) do
+    Kaffe.Producer.produce_sync(topic, [{key, value}])
+  end
+
+  def send_my_message(key, value) do
+    Kaffe.Producer.produce_sync(key, value)
+  end
+
+  def send_my_message(value) do
+    Kaffe.Producer.produce_sync("sample_key", value)
+  end
+end
+```
+
+**6. Send and receive messages in the console!**
+Start an interactive elixir shell with `iex -S mix` and call the following:
```

Comment:
So here we are using our `Example.Publisher` to publish to a queue that our example consumer is subscribing too? And that's why we see the message inspected to the console? Could we add a line here to sum that up/break it down? Its super cool to see it all come together!